File: lib/spiral/threading_contract.ts
import { defaultOf, IDisposable } from "../../../polyglot/deps/Fable/src/fable-library-ts/Util.js";
import { union_type, class_type, TypeInfo } from "../../../polyglot/deps/Fable/src/fable-library-ts/Reflection.js";
import { FSharpRef, Union } from "../../../polyglot/deps/Fable/src/fable-library-ts/Types.js";
import { value as value_1, some, Option } from "../../../polyglot/deps/Fable/src/fable-library-ts/Option.js";

export class Disposable implements IDisposable {
    readonly f: (() => void);
    constructor(f: (() => void)) {
        this.f = f;
    }
    Dispose(): void {
        const _: Disposable = this;
        _.f();
    }
}

export function Disposable_$reflection(): TypeInfo {
    return class_type("Threading.Disposable", undefined, Disposable);
}

export function Disposable_$ctor_3A5B6456(f: (() => void)): Disposable {
    return new Disposable(f);
}

export type US0_$union = 
    | US0<0>
    | US0<1>

export type US0_$cases = {
    0: ["US0_0", [any]],
    1: ["US0_1", []]
}

export function US0_US0_0(f0_0: any) {
    return new US0<0>(0, [f0_0]);
}

export function US0_US0_1() {
    return new US0<1>(1, []);
}

export class US0<Tag extends keyof US0_$cases> extends Union<Tag, US0_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US0_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US0_0", "US0_1"];
    }
}

export function US0_$reflection(): TypeInfo {
    return union_type("Threading.US0", [], US0, () => [[["f0_0", class_type("System.Threading.CancellationToken")]], []]);
}

export function US0__get_IsUS0_0(this$: FSharpRef<US0_$union>, unitArg: void): boolean {
    if ((this$ as any)['tag'] === 0) {
        return true;
    }
    else {
        return false;
    }
}

export function US0__get_IsUS0_1(this$: FSharpRef<US0_$union>, unitArg: void): boolean {
    if ((this$ as any)['tag'] === 1) {
        return true;
    }
    else {
        return false;
    }
}

export function closure1(unitVar: void, v0_1: any): US0_$union {
    return US0_US0_0(v0_1);
}

export function method0(): ((arg0: any) => US0_$union) {
    return (v: any): US0_$union => closure1(undefined, v);
}

export function closure2(v0_1: any, unitVar: void): void {
    let _v1: Option<void> = undefined;
    defaultOf();
    _v1 = some(undefined);
    value_1(_v1);
}

export function method1(v0_1: any): (() => void) {
    return (): void => {
        closure2(v0_1, undefined);
    };
}

export function closure0(unitVar: void, v0_1: Option<any>): [any, IDisposable] {
    let _v1: Option<[any, IDisposable]> = undefined;
    const patternInput: [any, IDisposable] = defaultOf();
    _v1 = ([patternInput[0], patternInput[1]] as [any, IDisposable]);
    const patternInput_1: [any, IDisposable] = value_1(_v1);
    return [patternInput_1[0], patternInput_1[1]] as [any, IDisposable];
}

export const v0 = (v: Option<any>): [any, IDisposable] => closure0(undefined, v);

export function new_disposable_token(x: Option<any>): [any, IDisposable] {
    return v0(x);
}

File: lib/spiral/threading__contract.py
from __future__ import annotations
from collections.abc import Callable
from typing import Any
from fable_modules.fable_library.async_builder import CancellationToken
from fable_modules.fable_library.option import (some, value as value_1)
from fable_modules.fable_library.reflection import (TypeInfo, class_type, union_type)
from fable_modules.fable_library.types import (Array, Union, FSharpRef)
from fable_modules.fable_library.util import IDisposable

def _expr65() -> TypeInfo:
    return class_type("Threading.Disposable", None, Disposable)


class Disposable(IDisposable):
    def __init__(self, f: Callable[[], None]) -> None:
        self.f: Callable[[], None] = f

    def Dispose(self, __unit: None=None) -> None:
        _: Disposable = self
        _.f(None)


Disposable_reflection = _expr65

def Disposable__ctor_3A5B6456(f: Callable[[], None]) -> Disposable:
    return Disposable(f)


def _expr66() -> TypeInfo:
    return union_type("Threading.US0", [], US0, lambda: [[("f0_0", class_type("System.Threading.CancellationToken"))], []])


class US0(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US0_0", "US0_1"]


US0_reflection = _expr66

def US0__get_IsUS0_0(this: FSharpRef[US0], unit_arg: None) -> bool:
    if this.tag == 0:
        return True

    else: 
        return False



def US0__get_IsUS0_1(this: FSharpRef[US0], unit_arg: None) -> bool:
    if this.tag == 1:
        return True

    else: 
        return False



def closure1(unit_var: None, v0_1: CancellationToken) -> US0:
    return US0(0, v0_1)


def method0(__unit: None=None) -> Callable[[CancellationToken], US0]:
    def _arrow67(v: CancellationToken) -> US0:
        return closure1(None, v)

    return _arrow67


def closure2(v0_1: Any, unit_var: None) -> None:
    _v1: None | None = None
    _v1 = some(None)
    value_1(_v1)


def method1(v0_1: Any) -> Callable[[], None]:
    def _arrow68(__unit: None=None, v0_1: Any=v0_1) -> None:
        closure2(v0_1, None)

    return _arrow68


def closure0(unit_var: None, v0_1: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    _v1: tuple[CancellationToken, IDisposable] | None = None
    pattern_input: tuple[CancellationToken, IDisposable] = None
    _v1 = (pattern_input[0], pattern_input[1])
    pattern_input_1: tuple[CancellationToken, IDisposable] = value_1(_v1)
    return (pattern_input_1[0], pattern_input_1[1])


def _arrow69(v: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    return closure0(None, v)


v0: Callable[[CancellationToken | None], tuple[CancellationToken, IDisposable]] = _arrow69

def new_disposable_token(x: CancellationToken | None=None) -> tuple[CancellationToken, IDisposable]:
    return v0(x)


